Add support for SW blink

fiam opened the pull request:
Allows displays which don't support blinking or with faulty blink implementations (like some MAX7456) to. Each display drivers declares the capabilities it supports and the display system handles blinking for the drivers that don't support it.

To workaround HW issues in MAX7456, a new setting has been introduced which makes the MAX7456 driver report its capabilities without support for blinking, so SW blink takes over.

Fixes #2419

fiam committed:
Remove blinkState global variable from osd.c

Not used anymore now what we support text attributes rather

fiam committed:
Add support for SW blink

Useful for chips which can't handle HW blink correctly or future
drivers that don't support HW blink at all.

Added a new setting named display_force_sw_blink which forces
drivers which support HW blink to use SW blink. For now the SW
blink cycle is 1sec on/1sec off to make it more evident while
testing.

fiam committed:
Make SW blink toggle the characters each 200ms

digitalentity commented on src/main/drivers/display.c:
```diff
@@ -85,6 +97,16 @@ int displayWrite(displayPort_t *instance, uint8_t x, uint8_t y, const char *s)
+    if (attr & ~instance->supportedTextAttributes) {
+        // We can't overwrite s, so we write char by char
+        int ret;
+        for (unsigned ii = 0; ii < strlen(s); ii++) {
```
This is suboptimal. `strlen` isn't declared `pure` or `const`, so gcc will evaluate it each loop iteration.

fiam replied:
Accidental `O(n^2)` ;-)

digitalentity replied:
Wel, not necesarily. If `strlen` is declared `pure` gcc _might_ pre-compute it since `s` is const. But I wouldn't count on that.

digitalentity commented on src/main/drivers/display.c:
```diff
@@ -85,6 +97,16 @@ int displayWrite(displayPort_t *instance, uint8_t x, uint8_t y, const char *s)
 int displayWriteWithAttr(displayPort_t *instance, uint8_t x, uint8_t y, const char *s, textAttributes_t attr)
 {
+    if (attr & ~instance->supportedTextAttributes) {
+        // We can't overwrite s, so we write char by char
```
Can't we reserve a buffer large enough and process the string in chunks? We fill the buffer, once it's filled - we flush it and continue with the rest of `s`.

digitalentity commented on src/main/io/displayport_max7456.c:
```diff
@@ -125,6 +125,19 @@ static uint32_t txBytesFree(const displayPort_t *displayPort)
+    textAttributes_t attr = TEXT_ATTRIBUTES_NONE;
+    TEXT_ATTRIBUTES_ADD_INVERTED(attr);
+    TEXT_ATTRIBUTES_ADD_SOLID_BG(attr);
+    if (!displayConfig()->force_sw_blink) {
```
In general drivers shouldn't use PGs. We need to pass the required configuration to `max7456Init` via `vcdProfile_t`.

fiam committed:
Move test for displayConfig()->force_sw_blink to displayInit()

This way we don't need to check for force_sw_blink in every driver,
the display system takes care of it.

fiam committed:
Use a local buffer in displayWriteWithAttr() for attribute emulation

This avoids calling into displayWriteCharWithAttr() for each
character. Note that support for arbitrary length string is disabled
behind a compile time option because it takes ~120 bytes of flash
to support it and right now the biggest screen width we support is
30 characters. Instead, we use a buffer which can hold up to 30
characters to perform any required rewrites.

digitalentity commented on src/main/io/displayport_oled.c:
```diff
@@ -109,7 +109,8 @@ static const displayPortVTable_t oledVTable = {
     .resync = oledResync,
-    .txBytesFree = oledTxBytesFree
+    .txBytesFree = oledTxBytesFree,
+    .supportedTextAttributes = NULL,
```
Why `NULL`? This is not a pointer.

fiam replied:
It is, the display driver vtable contains a `supportedTextAttributes` function pointer but the display caches it in a field. We can make the drivers change the field themselves before calling `displayInit()` instead, but it will be a bit less flexible. It saves 16 bytes of flash.

digitalentity replied:
Oh, ok, the cached value field name is the same which is confusing. How about refactoring function pointers in the vtable to have a `Fn` suffix?

fiam replied:
I think that's a bit too much for this PR. What about renaming the field to `cachedSupportedTextAttributes`?

digitalentity replied:
Fine for me

digitalentity commented on src/main/drivers/display.c:
```diff
@@ -85,8 +134,43 @@ int displayWrite(displayPort_t *instance, uint8_t x, uint8_t y, const char *s)
+        if (displayEmulateTextAttributes(instance, buf, s, blockSize, attr, &blockAttr)) {
+            // Emulation requires rewriting the string, which might be bigger than our buffer.
+            int ret = instance->vTable->writeString(instance, x, y, buf, blockAttr);
+#ifdef DISPLAY_ATTRIBUTE_EMULATION_MULTIPLE_WRITES
```
I'm not sure that `DISPLAY_ATTRIBUTE_EMULATION_MULTIPLE_WRITES` will be ever used.

fiam replied:
We can delete that, I guess we can always recover it from the history if we ever need it.

digitalentity replied:
Ok

fiam replied:
By deleting it we can also simplify `displayEmulateTextAttributes()` since it can now just overwrite the received attributes rather than copying them, saving 32 bytes of flash. Let's go for it.

digitalentity reviewed (commented):
Looks ok

fiam committed:
Remove code for emulating text attributes in multiple writes

Saves 32 bytes of flash, since we can simplify
displayEmulateTextAttributes().

fiam committed:
Rename field for text attr in displayPort_t

supportedTextAttributes -> cachedSupportedTextAttributes

Cosmetic change. Having the same name for the caching field than
the function pointer in the vtable could be confusing.
